### PR TITLE
Add Vial support for Keebio Sinc revs 1-3

### PR DESCRIPTION
**Pull request by kevinmost**
Based this off the initial QMK commit that added rev 3 support: https://github.com/qmk/qmk_firmware/tree/42c240f6cfe697744cdf876ace1e7273f43db60d/keyboards/keebio/sinc

Later snapshots of that directory would not work with Vial due to breaking changes in QMK that have not been merged.

Disabled quite a few features in revs 1 and 2 due to size constraints. However, rev 3 uses a RP2040 with much more flash memory, so nothing had to be disabled there.

Tested:
  - `make keebio/sinc/rev1:vial`, `make keebio/sinc/rev2:vial`, and `make keebio/sinc/rev3:vial` all compile successfully
  - Flashed the rev3 firmware to my Sinc rev3 and am able to use Vial

### CODE REVIEW

**Review by xyzz — CHANGES_REQUESTED** [review]
> Disabled quite a few features in revs 1 and 2 due to size constraints

these changes should go into the vial keymap, not into the revision. you can make a new keymap under e.g. keyboards/keebio/sinc/rev1/keymaps/vial - these all should also have different VIAL_KEYBOARD_UID since it's different pcb

**Comment by kevinmost** [comment]
@xyzz Thanks for the review, that makes sense. I updated the PR.

2 questions:

1) I had to put the `#define DYNAMIC_KEYMAP_LAYER_COUNT 0` definitions into `rev1/config.h` and `rev2/config.h`. Trying to put them in `rev1/keymaps/vial/config.h`, for example, resulted in the following error. Is this right?

```
./keyboards/keebio/sinc/rev2/keymaps/vial/config.h:28:0: error: "DYNAMIC_KEYMAP_LAYER_COUNT" redefined [-Werror]
 #define DYNAMIC_KEYMAP_LAYER_COUNT 1
 ^
In file included from <command-line>:0:0:
./.build/obj_keebio_sinc_rev2/src/info_config.h:45:0: note: this is the location of the previous definition
 #   define DYNAMIC_KEYMAP_LAYER_COUNT 3
 ^
```

2) Is a Vial firmware compiled with `DYNAMIC_KEYMAP_LAYER_COUNT` set to 0 actually functional, or does it mean that you don't even get a base layer to map? I did everything else to [reduce firmware size](https://get.vial.today/docs/firmware-size.html) and couldn't make the firmware fit. Even setting the layer count to 1 is not enough.

If this is non-functional, then should I just remove the rev1/rev2 directories and have this PR just add rev3 support?

(I only have a Sinc rev3, so I can't actually test the rev1 and rev2 firmwares on a device)

**Comment by lesshonor** [comment]
>     1. I had to put the `#define DYNAMIC_KEYMAP_LAYER_COUNT 0` definitions into `rev1/config.h` and `rev2/config.h`. Trying to put them in `rev1/keymaps/vial/config.h`, for example, resulted in the following error. Is this right?
> ```
> ./keyboards/keebio/sinc/rev2/keymaps/vial/config.h:28:0: error: "DYNAMIC_KEYMAP_LAYER_COUNT" redefined [-Werror]
>  #define DYNAMIC_KEYMAP_LAYER_COUNT 1
>  ^
> In file included from <command-line>:0:0:
> ./.build/obj_keebio_sinc_rev2/src/info_config.h:45:0: note: this is the location of the previous definition
>  #   define DYNAMIC_KEYMAP_LAYER_COUNT 3
>  ^
> ```

If you're redefining a preprocessor macro, you have to `#undef` it first.

e.g.:
https://github.com/vial-kb/vial-qmk/blob/47721416a6483a1d759c21a18beae3c0cb7d3eb2/keyboards/binepad/bnr1/keymaps/vial/config.h#L14-L17

**Comment by kevinmost** [comment]
That worked, thanks!

File: keyboards/keebio/sinc/info.json [hunk 1]
```diff
@@ -5,6 +5,22 @@
     "usb": {
         "vid": "0xCB10"
     },
+    "features": {
```
Comment (lesshonor) [inline keyboards/keebio/sinc/info.json:8]: Please revert all changes to the keyboard-level files. Per xyz it's not so bad if the changes match QMK but some of these definitely don't (upstream still has `sinc/rev1/config.h`)

You're welcome to do this in your own copy of the repository, but incorporating these changes now will make merging changes from upstream QMK more of a pain than it already is.

File: keyboards/keebio/sinc/keymaps/via/keymap.c [hunk 2]
```diff
@@ -37,3 +37,12 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______, _______, _______, _______, _______, _______
```
Comment (lesshonor) [inline keyboards/keebio/sinc/keymaps/via/keymap.c]: Please revert. Compiling VIA keymaps is not supported in the Vial repository (don't take my word for it; try running `make keebio/sinc:via` for yourself) so there is no need to modify anything about the `via` keymap.

File: keyboards/keebio/sinc/rev1/keymaps/vial/config.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* ws2812 RGB LED */
+#define RGBLIGHT_LED_MAP { 1, 2, 3, 12, 13, 14, 15, 0, 7, 6, 5, 4, 11, 10, 9, 8 }
+#define RGBLIGHT_DEFAULT_MODE (RGBLIGHT_MODE_RAINBOW_SWIRL + 2)
+
+// Set right column, left macro, 1.75u shift, split spaces, and 5x1u on by default
+#define VIA_EEPROM_LAYOUT_OPTIONS_DEFAULT 0x00DE
+#define VIA_EEPROM_LAYOUT_OPTIONS_SIZE 2
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev1/keymaps/vial/config.h]: Do you know that this works the way you expect? VIA ≠ Vial.

File: keyboards/keebio/sinc/rev1/keymaps/vial/config.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+// Esc + Enter. See: http://www.keyboard-layout-editor.com/#/gists/07a30619c76ab7d7be4932f94b3a7c06
+#define VIAL_UNLOCK_COMBO_ROWS { 5, 8 }
+#define VIAL_UNLOCK_COMBO_COLS { 2, 7 }
+
+// Reducing firmware size: https://get.vial.today/docs/firmware-size.html
+#ifdef DYNAMIC_KEYMAP_LAYER_COUNT
+    #undef DYNAMIC_KEYMAP_LAYER_COUNT
+#endif
+#define DYNAMIC_KEYMAP_LAYER_COUNT 0
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev1/keymaps/vial/config.h]:
```suggestion
```
> Is a Vial firmware compiled with DYNAMIC_KEYMAP_LAYER_COUNT set to 0 actually functional, or does it mean that you don't even get a base layer to map?

...You should understand the effects a setting will have before inflicting it on other people. Aside from doing something like flashing a test Pro Micro, you can most definitely flash your rev3 with this `#define` set, and see if you like the results.

By the way, [there are many other things that can be done to control size on AVR besides reducing the number of layers](https://docs.qmk.fm/#/squeezing_avr). Removing most RGB animations is very common. No one has mechanical locking keys either:

https://github.com/vial-kb/vial-qmk/blob/47721416a6483a1d759c21a18beae3c0cb7d3eb2/keyboards/mechwild/murphpad/keymaps/vial/config.h#L17-L39

Ultimately, if you don't want to go the extra mile it's fine if you just submit for rev3.

File: keyboards/keebio/sinc/rev3/keymaps/vial/vial.json [hunk 5]
```diff
@@ -0,0 +1,375 @@
+{
+  "name": "Sinc Rev. 3",
+  "vendorId": "0xCB10",
+  "productId": "0x3267",
+  "keycodes": ["qmk_lighting"],
+  "menus": ["qmk_rgb_matrix"],
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev3/keymaps/vial/vial.json]:
```suggestion
  "lighting": "vialrgb",
```
VIA ≠ Vial. Please reference [the Vial lighting documentation](https://get.vial.today/docs/lighting.html#rgb-matrix--vialrgb).

File: keyboards/keebio/sinc/rev3/keymaps/vial/rules.mk [hunk 6]
```diff
@@ -0,0 +1,5 @@
+VIA_ENABLE = yes
+VIAL_ENABLE = yes
+LTO_ENABLE = yes
+
+ENCODER_MAP_ENABLE = yes
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev3/keymaps/vial/rules.mk:5]:
```suggestion
ENCODER_MAP_ENABLE = yes

VIALRGB_ENABLE = yes
```
If you've got RGB_MATRIX and plenty of RAM, you might as well leverage Vial's UI.

**Review by lesshonor — CHANGES_REQUESTED** [review]

File: keyboards/keebio/sinc/info.json (hunk 1 above)
Comment (kevinmost) [reply keyboards/keebio/sinc/info.json:8]: The [current `vial-qmk`](https://github.com/vial-kb/vial-qmk/tree/vial/keyboards/keebio/sinc) main branch doesn't have the most recent `rev3` changes that are in the [upstream](https://github.com/qmk/qmk_firmware/tree/master/keyboards/keebio/sinc).

So I pulled in the current QMK upstream for the whole `keyboards/keebio/sinc` directory, and then added the `{rev1|rev2|rev3}/keymaps/vial` directories.

(Note: I did not merge in the `keymaps/` directory from upstream, but could do that too)

LMK if you would prefer that I do something else here, but this seems to make merging from upstream cleaner in the future, IMO, as the diff vs upstream is strictly new files now.

File: keyboards/keebio/sinc/keymaps/via/keymap.c (hunk 2 above)
Comment (kevinmost) [reply keyboards/keebio/sinc/keymaps/via/keymap.c]: Reverted.

File: keyboards/keebio/sinc/rev1/keymaps/vial/config.h (hunk 3 above)
Comment (kevinmost) [reply keyboards/keebio/sinc/rev1/keymaps/vial/config.h]: Reverted.

File: keyboards/keebio/sinc/rev1/keymaps/vial/config.h (hunk 4 above)
Comment (kevinmost) [reply keyboards/keebio/sinc/rev1/keymaps/vial/config.h]: Thanks for that link, that helps.

I was able to get rev1 and rev2 to compile by disabling:

- Locking keys
- Space Cadet
- Grave Escape

And I removed the changes to `DYNAMIC_KEYMAP_LAYER_COUNT`.

I do not have a Pro Micro to test on, unfortunately.

Would you prefer I revert any changes to the rev1 and rev2 directories? I can do that, or add some note about how this was not tested on a board.

Comment (kevinmost) [reply keyboards/keebio/sinc/rev1/keymaps/vial/config.h]: Side-note: Is it possible to add a link to that "Squeezing AVR" page to https://github.com/vial-kb/vial-kb.github.io/blob/main/docs/docs/firmware-size.md? I can send a PR.

File: keyboards/keebio/sinc/rev3/keymaps/vial/vial.json (hunk 5 above)
Comment (kevinmost) [reply keyboards/keebio/sinc/rev3/keymaps/vial/vial.json]: Done.

File: keyboards/keebio/sinc/rev3/keymaps/vial/rules.mk (hunk 6 above)
Comment (kevinmost) [reply keyboards/keebio/sinc/rev3/keymaps/vial/rules.mk:5]: Done.

**Comment by kevinmost** [comment]
Thanks for the review. I've done the following:

- Pulled in the latest [upstream](https://github.com/qmk/qmk_firmware/tree/master/keyboards/keebio/sinc) `sinc` directory as a base (which adds rev3 support)
- Ensured my edits on top of that are only in:
  - the new `keymaps/vial` directories for each revision
  - EDIT: Also the `{rev1|rev2|rev3}/config.h` (adding older RGB variable definitions for rev3, as well as the encoder pins for all revs)

**Review by kevinmost — COMMENTED** [review]
Hey, just pinging for a re-review when either of you gets a chance. Thanks.

File: keyboards/keebio/sinc/rev1/config.h [hunk 7]
```diff
@@ -18,45 +16,24 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 
-// Set right column, left macro, 1.75u shift, split spaces, and 5x1u on by default
-#define VIA_EEPROM_LAYOUT_OPTIONS_DEFAULT 0x00DE
 #define VIA_EEPROM_LAYOUT_OPTIONS_SIZE 2
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
+
+#define ENCODERS_PAD_A { F5 }
+#define ENCODERS_PAD_B { F6 }
+#define ENCODERS_PAD_A_RIGHT { D6 }
+#define ENCODERS_PAD_B_RIGHT { D4 }
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev1/config.h:39]:
```suggestion
```
If these are being moved to `info.json`, they don't need to be here.

File: keyboards/keebio/sinc/rev2/config.h [hunk 8]
```diff
@@ -16,45 +16,24 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
 
-// Set right column, left macro, 1.75u shift, split spaces, and 5x1u on by default
-#define VIA_EEPROM_LAYOUT_OPTIONS_DEFAULT 0x00DE
 #define VIA_EEPROM_LAYOUT_OPTIONS_SIZE 2
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
+
+#define ENCODERS_PAD_A { F5 }
+#define ENCODERS_PAD_B { F6 }
+#define ENCODERS_PAD_A_RIGHT { D6 }
+#define ENCODERS_PAD_B_RIGHT { D4 }
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev2/config.h:39]:
```suggestion
```

File: keyboards/keebio/sinc/rev3/config.h [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#define DRIVER_LED_TOTAL RGB_MATRIX_LED_COUNT
+#define RGBLED_NUM RGB_MATRIX_LED_COUNT
+#define RGBLIGHT_SPLIT
+#define RGBLED_SPLIT { 56, 57 }
+#define RGB_MATRIX_SPLIT RGBLED_SPLIT
+
+#define ENCODERS_PAD_A { GP20 }
+#define ENCODERS_PAD_B { GP21 }
+#define ENCODERS_PAD_A_RIGHT { GP6 }
+#define ENCODERS_PAD_B_RIGHT { GP5 }
```
Comment (lesshonor) [inline keyboards/keebio/sinc/rev3/config.h:89]:
```suggestion
```

**Review by lesshonor — CHANGES_REQUESTED** [review]
...so, one wrench in the gears is that at this point [there's a pending branch to pull in relatively recent QMK updates](https://github.com/vial-kb/vial-qmk/tree/merge-2023-03-12).

I have no control over pull requests, but I wouldn't expect this to be merged until after those changes are brought in. That way the changes you had to make to cover the 0.18.x to 0.20.x gap don't have to be dealt with later, and it's simpler to examine PRs for Vial keymaps that are *just* Vial keymaps.

One option might be retargeting your PR to that branch; it presumably wouldn't have to incorporate any keyboard-level changes that way.
(If you'd like, [you can also betatest the new version of the GUI](https://github.com/vial-kb/vial-gui/actions/runs/4463584822).)

**Pull request closed without merging**